### PR TITLE
Added APD Id to path

### PR DESCRIPTION
**Pull request by thetif**
issue #2589

APD paths should include the eAPD id so that it's easier to reference an APD

### This pull request changes...

- the apdId is now part of the url for all APD urls

### This is how to verify this change...

- verify that all APD urls now contain the apdId
- verify that all of the navigation buttons and links still work
- verify that copying an APD url and pasting it in another window should open the APD to the correct page
- verify that copying the print url and pasting it in another window should open the APD to the correct page

### This pull request is ready to merge when...

- [X] Develop has updated automated tests (and all tests are passing)
- [x] This code has been reviewed by someone other than the original author
- [ ] QA has manually tested and created bugs for any issues found (beyond minor fixes)
- [X] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented
- [X] The change has been documented
~- [ ] Associated OpenAPI documentation has been updated~
- [X] Changelog is updated as appropriate

### This feature is done when...

- [ ] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

**Review by knollfear — APPROVED** [review]
LGTM!

**Review by tbolt — APPROVED** [review]

**Comment by beparticular** [comment]
@thetif @eStar386  I'm seeing issues with the copy/pasting URLs into new tabs.
I'm using Chrome. 

NOT WORKING AS EXPECTED, consistently
- verify that copying an APD url and pasting it in another window should open the APD to the correct page
When I try to copy/paste an APD url into another window, it looks like I'm being taken to the login, I'm getting a notice that my session is being validated, a notice that the APD is loading, then I land on the right APD, but at the start of the APD. 
This is happening consistently, and I haven't gotten the expected behavior in several tries now, with several different page links. Here are some examples
https://ec2-34-193-194-48.compute-1.amazonaws.com/apd/3/export
https://ec2-34-193-194-48.compute-1.amazonaws.com/apd/2/state-profile
https://ec2-34-193-194-48.compute-1.amazonaws.com/apd/2/activity/1/oms

WORKING AS EXPECTED, SOMETIMES? Maybe all the time? 
- verify that copying the print url and pasting it in another window should open the APD to the correct page
This one, I'm getting different results with different links so I'll report as best I can. 

SOMETIMES _ When I try to copy/paste the print URL, I'm getting a notice that my session is being validated, a notice that the APD is loading, then I'm taken to the right APD, but to the start of the APD. I know I saw this happen, but I'm wondering if I pasted in the https://ec2-34-193-194-48.compute-1.amazonaws.com/apd/3/export link instead of the print link and got the behavior described above because I can't reproduce it when I'm trying to reproduce it. 

I've copy/pasted the print URL successfully into a new window many times. Using this link https://ec2-34-193-194-48.compute-1.amazonaws.com/print/1 and this one https://ec2-34-193-194-48.compute-1.amazonaws.com/print/2 as well. Consistently. So, the first time I did this might have been a fluke, but is worth checking to see if anyone else can make it happen. 

I tried to see if it was related to other links or APDs being open, but that doesn't seem to be an issue. 

Separately - I tried to log out of this test environment and was unable to the first time. I haven't been able to reproduce this either, so maybe whatever I saw on my first attempt at testing the print link was an issue with the test link, or if there are issues here. 

Let me know if y'all have any questions.

**Comment by eStar386** [comment]
@beparticular I have noticed the same thing and have been trying to figure out if i provide any more details than steps to reproduce. As stated, initially I didn't notice it but the only thing I can add is that it could be cookie based side effect causing the issue because initially it worked but then it displays the aforementioned issues.

**Comment by thetif** [comment]
very strange, I will pull this back into in progress and look at it again.

**Comment by thetif** [comment]
@beparticular and @eStar386 I think I resolved this. It ended up being something pretty simple. I'm going to finish testing it tomorrow and if it seems to fix all those issues above I'll send it back to Code review and QA.

**Comment by thetif** [comment]
@beparticular and @eStar386 I have tested the above instances and they should be working now. Please test again.

**Comment by beparticular** [comment]
@thetif one last question on this, is the intended behavior for the top level nav to work as a sharable link, but not the sub navs? For instance, if I click on Actual Expenditures in the side nav and share that link, the shared link redirects to the top level Results of Previous Activities nav.

**Comment by eStar386** [comment]
Minor Functional Issue:When creating and editing activities. It starts at 0 in the url however the actual activity starts at one. For consistency and easy readability the side nav/activity url should be the same.  
[urls_2589.pdf](https://github.com/CMSgov/eAPD/files/6481015/urls_2589.pdf)

**Comment by beparticular** [comment]
Per parking lot, let's add the URL find to a new issue and assign to design to consider the human needs and behaviors around sharing URLs.

**Comment by knollfear** [comment]
I found some weirdness in testing.  The link on the home page that takes you to '<website>/apd/62/apd-overview' shows as just "<website>/#!" when you copy the link.  I was also able to fetch an APD that was archived by just adding the number to the URL.  Finally, I got myself stuck when I tried to go to an APD that did not exist by changing the address to '<website>/apd/999/apd-overview'

File: web/src/containers/ApdApplication.js [hunk 1]
```diff
@@ -1,34 +1,72 @@
...
+  useHistory,
+  useLocation
 }) => {
+  const [isLoading, setIsLoading] = useState(false);
+  const paramApdId = +useParams().apdId;
+  const history = useHistory();
+  const location = useLocation();
+
+  useEffect(() => {
+    console.log({ location });
```
Comment (tbolt) [inline web/src/containers/ApdApplication.js:40]: Probably take this out

**Review by tbolt — APPROVED** [review]

File: web/src/containers/ApdApplication.js (hunk 1 above)
Comment (thetif) [reply web/src/containers/ApdApplication.js:40]: got it

**Comment by thetif** [comment]
Discussed an issue with the /apds/id endpoint allowing all types of APDs to be returned, we agreed that for the time being we would return only APDs that could be edited.

**Comment by thetif** [comment]
discussed issue of how to handle the user interaction if they tried to access an invalid url with @beparticular 

> I think you could return them to the dashboard with an alert showing from the design system that says “This APD does not exist at this link. Check your link to make sure it is correct.” (or something like that).

File: web/src/components/ApdList.js [hunk 2]
```diff
@@ -68,6 +60,14 @@ const ApdList = (
         <main id="start-main-content">
           <div className="ds-u-padding-top--2">
             <UpgradeBrowser />
+            {error && (
+              <div className="ds-c-alert ds-c-alert--error">
```
Comment (tbolt) [inline web/src/components/ApdList.js:64]: You could use the react component version of this `<Alert>` from design.cms.gov here. 
https://design.cms.gov/components/alert/

**Review by tbolt — APPROVED** [review]
Looks good. Made a small suggestion but not critical.

**Comment by eStar386** [comment]
Regression Testing of Back and Forth Features shows some differences. Please see and verify this is the correct functionality. 

Uploading Screen Recording 2021-05-24 at 4.00.01 PM.mov…

**Comment by thetif** [comment]
@eStar386 I've been looking into this for a couple of hours and I am unable to resolve it. I can definitely reproduce it though. Single Page Applications like this one are pretty hard to keep the back and forward buttons working as expected because the routing is virtual unlike on Multiple Page Application where the url actually changes on every page change. I looked and this is already happening in the application on staging. Also, that's why we put the back buttons on the page so we can control the back functionality more than with the browser buttons. I think it is an odd idiosyncrasy of working with a react router. I think we should log this a bug to investigate later, but fail the ticket for it. Does that work for you?

**Review by eStar386 — CHANGES_REQUESTED** [review]
https://docs.google.com/spreadsheets/d/1LqdcfO6-VUGs3pSpoGu8XTv0O9euktNjmKC7FHbKJ3Q/edit?usp=sharing  
Per the github discussion, there are some idiosyncrasies with the react router. @thetif @beparticular Please confirm expected functionality for regression of navigation buttons. 
 1. External links the user will use browser buttons.
 2. Internal the user will only use use the back and continue buttons located within the apd application.

**Comment by beparticular** [comment]
@eStar386 
We expect the users inside of the application to use the back and continue buttons provided in the application, as well as the “back to APD” button on the export view and not the browser buttons.

Was the change request for confirmation of that information, or was there something else as well?

If there are no other changes requested, please go ahead and move to ready to review, as this additional testing on this part of the issue should be considered in the bug 3161. Thanks!

**Review by eStar386 — APPROVED** [review]
Exemptions documented in the test cases.

**Review by beparticular — APPROVED** [review]
This all appears to be working swimmingly!

**Review by NAretakis — APPROVED** [review]
Good to go!

**Pull request merged**